edx_notifications/stores/mongo/store_provider.py

```diff
@@ -0,0 +1,141 @@
+
+        if not (read and unread):
+            if read:
+                query_object.update({'user_notification.read_at': {"$ne": None}})
+
+            if unread:
+                query_object.update({'user_notification.read_at': None})
+
+        if type_name:
+            query_object.update({'user_notification.msg_type_name': type_name})
+
+        if start_date:
+            query_object.update({'created': {"gte": start_date}})
+
+        if end_date:
+            query_object.update({'created': {"lte": end_date}})
+
+        collection = self.collection.find(query_object)
+        return list(collection)
+
+    def get_notifications_for_user(self, user_id, filters=None, options=None):
+        """
+
+        :param user_id:
+        :param filters:
+        :param options:
+        :return:
+        """
+        user_notifications = self._get_prepaged_notifications(user_id, filters, options)[0]
+
+        result_set = [MongoUserNotification.to_data_object(item, user_id) for item in user_notifications['user_notification']]
+
+        return result_set
+
+    def mark_user_notifications_read(self, user_id, filters=None):
+        pass
+
+    def get_num_notifications_for_user(self, user_id, filters=None):
+        return len(self._get_prepaged_notifications(user_id, filters)[0]['user_notification'])
+
+    def purge_expired_notifications(self, purge_read_messages_older_than, purge_unread_messages_older_than):
+        pass
+
+    def bulk_create_user_notification(self, user_msgs):
+        """
+        This is an optimization for bulk creating *new* UserNotification
+        objects in the mongodb . Since we want to support fan-outs of messages,
+        we may need to insert 10,000's (or 100,000's) of objects as optimized
+        as possible.
+
+        NOTE: This method cannot update existing UserNotifications, only create them.
+        NOTE: It is assumed that user_msgs is already chunked in an appropriate size.
+        """
+        if len(user_msgs) > const.NOTIFICATION_BULK_PUBLISH_CHUNK_SIZE:
+            msg = (
+                'You have passed in a user_msgs list of size {length} but the size '
+                'limit is {max}.'.format(length=len(user_msgs), max=const.NOTIFICATION_BULK_PUBLISH_CHUNK_SIZE)
+            )
+            raise BulkOperationTooLarge(msg)
+
+    def get_notification_for_user(self, user_id, msg_id):
+        collection = self.collection.find({
+            'user_id': user_id,
+            'user_notification.msg_id': msg_id
+        })
+        return list(collection)
+
+    def save_user_notification(self, user_msg):
+        """
+        Create or Update the mapping of a user to a notification.
+        """
+
+        user_notification = self.collection.update(
+            {
+                'user_id': user_msg.user_id,
+            },
+            {
+                '$set': {'user_id': user_msg.user_id},
+                '$push': {
```

Comment:
In terms of default ordering where we want the most recent notification first, can we insert the new user_notification at the beginning of the sub-collection array? Sorry, I'm not sure if $push here means 'append at the end' or 'insert at the beginning'.